### PR TITLE
Macro diagnostics tweaks

### LINKED ISSUE #10951: Macro with empty expansion in expression position causes an "unexpected <eof>" error
``` rust
macro_rules! foo ( () => () );
fn main() { foo!() }
```

```
expand.rs:1:0: 1:0 error: unexpected token: `<eof>`
expand.rs:1 macro_rules! foo ( () => () );
            ^
```

(Moving the `foo!()` outside `main` works.)

### LINKED ISSUE #26288: Confusing error message from missing semicolon after macro invocation
This code

```
macro_rules! foo {
    () => {
        assert_eq!("A", "A");
        assert_eq!("B", "B");
    }
}

fn main() {
    foo!()
}
```

gives the error:

```
foo.rs:4:9: 4:18 error: macro expansion ignores token `assert_eq` and any following
foo.rs:4         assert_eq!("B", "B");
                 ^~~~~~~~~
foo.rs:9:5: 9:11 note: caused by the macro expansion here; the usage of `foo` is likely invalid in this context
foo.rs:9     foo!()
             ^~~~~~
error: aborting due to previous error
```

The problem is actually a missing semicolon after `foo!()`, but the message wasn't really helpful in figuring that out.  (I initially left out the semicolon out of habit from C, where including it would result in an extra empty statement after expansion.)

### LINKED ISSUE #30128: macro expanding to nothing in match arm causes uninformative parse error
[Failing code](http://is.gd/q92sdB):

``` rust
macro_rules! empty { () => () }

fn main() {
    match 42 {
        _ => empty!()
    };
}
```

The error is "unexpected token: `<eof>`" with a span of 1:1: 1:1.

[Workaround](http://is.gd/93TwC5):

``` rust
macro_rules! empty { () => () }

fn main() {
    match 42 {
        _ => { empty!(); }
    };
}
```

Expected: the failing code either compiles or at least gives an error pointing at the problem.

Found by mohawk on IRC.

### PR DESCRIPTION
**Pull request by estebank**
Fix #30128, fix #10951 by adding an appropriate span to the diagnostic.
Fix #26288 by suggesting adding semicolon to macro call.

### CODE REVIEW

File: src/test/ui/macros/macro-in-expression-context.stderr [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+   |
+LL |         assert_eq!("B", "B");
+   |         ^^^^^^^^^
+...
+LL |     foo!()
+   |     ------- help: you might be missing a semicolon here: `;`
+   |     |
+   |     caused by the macro expansion here
+   |
+   = note: the usage of `foo!` is likely invalid in expression context
```
Comment (estebank) [inline src/test/ui/macros/macro-in-expression-context.stderr]: @dtolnay can you take a look and verify wether this suggestion is reasonable (it won't be wrong more often than not)?

Comment (estebank) [reply src/test/ui/macros/macro-in-expression-context.stderr]: This is one case where it'd be the wrong approach:

```
fn main() {
    macro_rules! a {
        ($e:expr) => { $e; 2}
    }
    a!(true);
}
```
```
error: macro expansion ignores token `2` and any following
 --> file.rs:3:28
  |
3 |         ($e:expr) => { $e; 2}
  |                            ^
4 |     }
5 |     a!(true)
  |     --------- help: you might be missing a semicolon here: `;`
  |     |
  |     caused by the macro expansion here
  |
  = note: the usage of `a!` is likely invalid in expression context
```

but the suggestion actually gets the code to compile...

File: src/test/ui/macros/macro-context.stderr [hunk 2]
```diff
@@ -3,36 +3,35 @@ error: macro expansion ignores token `;` and any following
...
 LL |     () => ( i ; typeof );   //~ ERROR expected expression, found reserved keyword `typeof`
    |                 ^^^^^^
-   |
-note: caused by the macro expansion here; the usage of `m!` is likely invalid in expression context
-  --> $DIR/macro-context.rs:21:13
-   |
+...
 LL |     let i = m!();
-   |             ^^^^
+   |             ----- help: you might be missing a semicolon here: `;`
```
Comment (pnkfelix) [inline src/test/ui/macros/macro-context.stderr]: Can we not peek ahead in the span and see if there's a semicolon as the next character before making such an obviously inapplicable help message?

**Comment by pnkfelix** [comment]
This seems fine. I would like an answer to my Q about peeking at the next character rather than blindly stating that a semicolon might be missing when it clearly is present in the source.

File: src/test/ui/macros/macro-context.stderr (hunk 2 above)
Comment (estebank) [reply src/test/ui/macros/macro-context.stderr]: [We're actually checking that](https://github.com/rust-lang/rust/pull/55292/files#diff-7fde1475846894e59317d75947bdb921R1048), which makes me intrigued at why this isn't being picked up. My guess is that if spans were completely hygienic this suggestion would appear on line 13, after `typeof`. This seems to be the only edge-case caught by our tests (before I added the linked check, many other incorrect suggestions were being emitted). Given that, would you mind if we merge as is and add extra logic in a later PR?

Comment (pnkfelix) [reply src/test/ui/macros/macro-context.stderr]: Sounds good

**Comment by estebank** [comment]
@bors r=pnkfelix

@pnkfelix went ahead and fixed it.

**Pull request merged**